docs/oppen_algorithm.md

```diff
@@ -0,0 +1,359 @@
+# Derek C. Oppen's pretty pritning algorithm
+
+Derek C. Oppen's algorithm provides an efficient solution for formatting and printing code with clean indentation that respects a specified screen width.
+
+When working with large blocks of code, ensuring proper indentation and alignment is crucial for readability and maintainability. However, achieving this while adhering to a specific line width can be a challenging task.
+
+This algorithm is great if you are working with code, documentation, logging, or any situation that requires pretty-printing.
+
+## The printing tokens
+
+Oppen's algorithm expects a list of tokens describing the structure of the expected output.
+
+The main tokens are:
+
+|   Name    | Description
+|-----------|------------
+| STRING    | String literal to be displayed.
+| BEGIN     | Opens a group.
+| END       | Closes a group.
+| BREAK     | Potential line break (adds a new line if the text doesn't fit on the line).
+| LINEBREAK | Guaranteed line break.
+
+A group can contain any number of groups, `String`, `Break`, and a `LineBreak`. Groups exist to containerize different _groups_ of tokens with respect to line breaking.
+
+Containerizing groups allows the creation of subsets that are independent from other subsets, and semi-independent from their parent subsets.
+
+Example for a maximum desired printing width of 3:
+```
+[
+  BEGIN#1, # Start of group 1
+
+  STRING = 'a',
+
+  BREAK#1,
+  STRING = 'b',
+
+  BREAK#2,
+  STRING = 'c',
+
+  END#1,   # end of group 1
+
+  EOF
+]
+
+=>
+a
+b
+c
+```
+Now if we add one more group:
+```
+[
+  BEGIN#1, # Start of group 1
+
+  STRING = 'a',
+
+  BEGIN#2, # Start of group 2
+
+  BREAK#1,
+  STRING = 'b',
+
+  END#2,   # end of group 2
+
+  BREAK#2,
+  STRING = 'c',
+
+  END#1,   # end of group 1
+
+  EOF
+]
+
+=>
+a b
+c
+```
```

Comment:
````suggestion
If we introduce a new group as follows:
```
[BEGIN, STRING(a), 
    BEGIN, BREAK, STRING(b), END,
 BREAK, STRING(c), END]
```

We get:
```
a b
c
```

Instead of replacing all the spaces introduced by `BREAK` by an actual new line to fit a line of width 3, this configuration allowed us to break the outermost group only while keeping the inner `BREAK` (preceding `STRING('b')`) intact.
````